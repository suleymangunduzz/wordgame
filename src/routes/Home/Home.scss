@import '../../styles/index';

.Home {
  display: flex;
  width: 100%;
  height: 100%;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  &__play-area {
    position: relative;
    display: flex;
    min-width: 20rem;
    min-height: 10rem;
    align-items: center;
    justify-content: center;
    border: 0.2rem solid rgb(7, 6, 6);
    margin-bottom: 2rem;
    border-radius: 1.2rem;
  }

  &__winner {
    color: $green-color;
  }
}
