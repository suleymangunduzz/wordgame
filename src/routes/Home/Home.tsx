import React, { FunctionComponent, useState, useEffect, useRef } from 'react';
import './Home.scss';

import Button from '../../components/Button';
import ResultPanel from '../../components/ResultPanel';
import PlayerInfo from '../../components/PlayerInfo';
import TimePanel from '../../components/TimePanel';
import Selections from '../../components/Selections';

import names from '../../utils/names.json';

import useInterval from '../../utils/useInterval';
import listenUser from '../../utils/ListenUser';
import listenMachine from '../../utils/ListenMachine';
import generateRandomNumber from '../../utils/RandomNumberGenerator';

import { NOT_STARTED, STARTED, FINISHED, USER, MACHINE } from '../../utils/constants';

const Home: FunctionComponent = () => {
  const emptyState: string[] = [];
  const [gameStatus, setGameStatus] = useState<string>(NOT_STARTED);
  const [selectedNames, setSelectedNames] = useState<string[]>([]);
  const [nextPlayer, setNextPlayer] = useState<string>('');
  const [winner, setWinner] = useState<string>('');
  const [timeLeft, setTimeLeft] = useState<number>(8);
  const [usersFinalAnswer, setUsersFinalAnswer] = useState<string>('');
  const synthRef = useRef(window.speechSynthesis);

  let globalTimer: any;
  let globalInterval: any;

  useEffect(() => {
    if (gameStatus === STARTED) {
      const clearNames = async () => {
        await setSelectedNames(emptyState);
      };

      clearNames();
    }
  }, [gameStatus]);

  useEffect(() => {
    if (timeLeft === 0 && nextPlayer === USER && usersFinalAnswer === '') {
      setNextPlayer(MACHINE);
      setGameStatus(FINISHED);
      setWinner(MACHINE);
    }
  }, [timeLeft, nextPlayer, usersFinalAnswer]);

  useInterval(() => setTimeLeft(timeLeft - 1), timeLeft, gameStatus);

  let recognition: any;

  if ('SpeechRecognition' in window || 'webkitSpeechRecognition' in window) {
    const SpeechRecognition =
      (window as any).SpeechRecognition || (window as any).webkitSpeechRecognition;

    recognition = new SpeechRecognition();
    recognition.continuous = true;
    recognition.interimResults = true;
    recognition.lang = 'tr-TR';
  }

  useEffect(() => {
    if (gameStatus !== FINISHED) {
      switch (nextPlayer) {
        case MACHINE:
          listenMachine(
            generateRandomNumber(10) > 0,
            generateRandomNumber(7),
            selectedNames,
            names,
            synthRef,
            setWinner,
            setTimeLeft,
            setGameStatus,
            setNextPlayer,
            setUsersFinalAnswer,
            setSelectedNames,
            false,
          );
          break;
        case USER:
          listenUser(
            recognition,
            selectedNames,
            names,
            setWinner,
            setTimeLeft,
            setGameStatus,
            setNextPlayer,
            setUsersFinalAnswer,
            setSelectedNames,
          );
          break;
        default:
          break;
      }
    }

    return () => {
      recognition.stop();
      clearTimeout(globalTimer);
      clearInterval(globalInterval);
    };
  }, [nextPlayer]);

  const startGameHandler = () => {
    if (gameStatus !== STARTED) {
      setGameStatus(STARTED);
      listenMachine(
        generateRandomNumber(10) > 0,
        generateRandomNumber(7),
        selectedNames,
        names,
        synthRef,
        setWinner,
        setTimeLeft,
        setGameStatus,
        setNextPlayer,
        setUsersFinalAnswer,
        setSelectedNames,
        true,
      );
      setTimeLeft(8);
    }
  };

  const renderGameStatus = () => {
    return gameStatus === FINISHED && <div className="Home__winner">{`Kazanan: ${winner}`}</div>;
  };

  const getStartButtonText = () => {
    return gameStatus !== STARTED ? 'Oyuna Ba≈üla' : 'Oyun Devam Ediyor';
  };

  return (
    <div className="Home">
      <h1>Kelime Oyunu</h1>

      <div className="Home__play-area">
        <ResultPanel gameStatus={gameStatus} selectedNames={selectedNames} />
        <PlayerInfo gameStatus={gameStatus} nextPlayer={nextPlayer} />
        <TimePanel timeLeft={timeLeft} gameStatus={gameStatus} />
      </div>

      <Button onClick={startGameHandler} disabled={gameStatus === STARTED}>
        {getStartButtonText()}
      </Button>

      {renderGameStatus()}

      <Selections gameStatus={gameStatus} selectedNames={selectedNames} />
    </div>
  );
};

export default Home;
