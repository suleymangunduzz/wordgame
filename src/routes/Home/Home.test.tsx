import React from 'react';
import { shallow } from 'enzyme';

import Home from './Home';

declare global {
  interface Global {
    navigator: any;
  }
}

const wrapper = shallow(<Home />);

it('renders correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

it('should start game correctly', () => {
  wrapper.find('Button').simulate('click');
});
