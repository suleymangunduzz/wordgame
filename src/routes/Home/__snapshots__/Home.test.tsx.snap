// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Home"
>
  <h1>
    Kelime Oyunu
  </h1>
  <div
    className="Home__play-area"
  >
    <ResultPanel
      gameStatus="NOT_STARTED"
      selectedNames={Array []}
    />
    <PlayerInfo
      gameStatus="NOT_STARTED"
      nextPlayer=""
    />
    <TimePanel
      gameStatus="NOT_STARTED"
      timeLeft={8}
    />
  </div>
  <Button
    disabled={false}
    onClick={[Function]}
  >
    Oyuna Ba≈üla
  </Button>
  <Selections
    gameStatus="NOT_STARTED"
    selectedNames={Array []}
  />
</div>
`;
